src/util/http_utils.js

```diff
@@ -5,15 +5,23 @@ const _ = require('lodash');
 const url = require('url');
 const http = require('http');
 const https = require('https');
+const HttpProxyAgent = require('http-proxy-agent');
+const HttpsProxyAgent = require('https-proxy-agent');
 const crypto = require('crypto');
 const xml2js = require('xml2js');
 const querystring = require('querystring');
+const ip = require('ip');
 const P = require('./promise');
 const dbg = require('./debug_module')(__filename);
 const xml_utils = require('./xml_utils');
 const cloud_utils = require('./cloud_utils');
 const util = require('util');
 
+
+const no_proxy_list = process.env.NO_PROXY ?
+    process.env.NO_PROXY.split(',') :
+    [];
```

Comment:
what do you think about mapping each item to a should_proxy() function based on it's type, something like:
```suggestion
/**
 * should_not_proxy_list is an array of functions that each check a given address for a no proxy match
 * @type Array<(address: string) => boolean>
 */
const should_not_proxy_list = (process.env.NO_PROXY || []).split(',').map(cond => {

    // Cover the case where cond is an IP
    if (ip.isV4Format(cond) || ip.isV6Format(cond)) return x => ip.isEqual(x, cond);

    // Cover the case where cond is a CIDR subnet (eg. 10.20.30.40/24)
    let subnet;
    try { subnet = ip.cidrSubnet(cond); } catch () {}
    if (subnet) return x => subnet.contains(x);

    // Cover the case where cond is domain name suffix
    if (cond.startsWith('.')) return x => x.endsWith(cond);

    // Cover the case where cond is an explicit FQDN
    return x => (x === addr);
});

function should_proxy(address) {
    // if all the "should not proxy" functions return false, we should proxy.
    // if any one of "should not proxy" functions returns true for address, we shouldn't proxy.
    return should_not_proxy_list.every(fn => !fn(address));
}

```

Reply:
I was also thinking about it,
But it will just create more complexity in a non-crutial code path. I prefer to provide a simpler solution until we will prove that we are bottlenecked on the addresses type check

Reply:
@guymguym Also, you have a bug in the proposed code. you cannot split on an empty array, and even if you choose to set the default value to an empty string it will create an array with one empty item and not an empty array.

Reply:
Changed the code a bit to pre-process the list immediately after the split. It is not exactly your suggestion but it allows us to select tests based on the hostname type (IP or FQDN) and does not invoke a multitude of functions.

Reply:
ok, resolving.